py/selenium/webdriver/common/bidi/network.py

```diff
@@ -0,0 +1,262 @@
+                        'credentials': {
+                            'type': 'password',
+                            'username': username,
+                            'password': password
+                        }
+                    }
+        }
+        self.conn.execute(self.command_iterator(command))
+
+    def __on(self, event, callback):
+        """Set a callback function to subscribe to a network event."""
+        event = self.EVENTS.get(event, event)
+        self.callbacks[event] = callback
+        if len(self.subscriptions[event]) == 0:
+            # session_subscribe(self.conn, event, self.__handle_event)
+            self.conn.add_callback(event, self.__handle_event)
+
+    def __handle_event(self, event, data):
+        """Perform callback function on event."""
+        if event in self.callbacks:
+            self.callbacks[event](data)
+
+    def add_authentication_handler(self, username, password):
+        """Adds an authentication handler."""
+        self.__add_intercept(phases=[self.PHASES['auth_required']])
+        self.__on('auth_required', lambda data: self.__continue_with_auth(data['request']['request'], username, password))
+        self.subscriptions['auth_required'] = [username, password]
+
+    def remove_authentication_handler(self):
+        """Removes an authentication handler."""
+        self.__remove_intercept(intercept='auth_required')
+        del self.subscriptions['auth_required']
+        session_unsubscribe(self.conn, self.EVENTS['auth_required'])
+
+    def add_request_handler(self, callback, url_pattern=''):
+        """Adds a request handler that executes a callback function when a
+        request matches the given URL pattern.
+
+        Parameters:
+            callback (function): A function to be executed when url is matched by a URL pattern
+                The callback function receives a `Response` object as its argument.
+            url_pattern (str, optional): A substring to match against the response URL.
+                Default is an empty string, which matches all URLs.
+
+        Returns:
+            str: The request ID of the intercepted response.
+        """
+        self.__add_intercept(phases=[self.PHASES['before_request']])
+        request_id = None
+        def callback_on_url_match(data):
+            nonlocal request_id
+            if url_pattern in data['request']['url']:
+                # create request object to pass to callback
+                request_id = data['request'].get('requestId', None)
+                url = data['request'].get('url')
+                method = data['request'].get('method')
+                headers = data['request'].get('headers', {})
+                body = data['request'].get('body', None)
+                request = Request(request_id, url, method, headers, body, self)
+                callback(request)
+        self.__on('before_request', callback_on_url_match)
+        self.callbacks[request_id] = callback
```

Comment:
I don't think it is ever going to work - it will always end up being `{None: callback}`. This way, the callback will never really be called. It's likely a key issue why your tests are failing.

Looking at Ruby implementation, you need to return callback ID rather than request ID:
https://github.com/SeleniumHQ/selenium/blob/b2cea1debe78b6554371987357ae1c54b30ec5bf/rb/lib/selenium/webdriver/common/network.rb#L61-L71